internal/surfer/gcloud/utils/resource.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/googleapis/librarian/internal/sidekick/api"
+	"github.com/iancoleman/strcase"
+)
+
+// GetPluralFromSegments infers the plural name of a resource from its structured path segments.
+// Per AIP-122, the plural is the literal segment before the final variable segment.
+// Example: `.../instances/{instance}` -> "instances".
+func GetPluralFromSegments(segments []api.PathSegment) string {
```

Comment:
I like how these helpers are abstracted by type! We want to make they don't become redundant or unwieldy